### PR TITLE
Austin/homefeed improvements

### PR DESCRIPTION
**Pull request by AAAstorga**
- Added search tab and all the functionality that comes along with it
- Realm fixes
- bookmark fixes
- fixed all old issues with old fat PR before break
- quality content

### CODE REVIEW

File: Cornell Sun/Cornell Sun/SearchViewController.swift [hunk 1]
```diff
@@ -7,81 +7,265 @@
...
         // Set up the searchController delegate and the presentation view
-        searchController.searchResultsUpdater = self
-        searchController.delegate = self
-        searchController.dimsBackgroundDuringPresentation = false
+        searchController.searchBar.delegate = self
         searchController.definesPresentationContext = true
-
-        // Could also make the middle element of the navbar equal to this
-        tableView.tableHeaderView = searchController.searchBar
+        searchController.dimsBackgroundDuringPresentation = false
```
Comment (mindylou) [inline Cornell Sun/Cornell Sun/SearchViewController.swift]: can we also have the search controller dismiss when you tap away from the search bar?

File: Cornell Sun/Cornell Sun/Cells/SearchCell.swift [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import UIKit
+import SnapKit
+
+final class SearchCell: UICollectionViewCell {
+
+    var insetConstant: CGFloat = 18
+    var offsetConstant: CGFloat = 12
+    var imageViewWidthHeight: CGFloat = 90
+    var timeLabelOffset: CGFloat = -8
```
Comment (cdsciavolino) [inline Cornell Sun/Cornell Sun/Cells/SearchCell.swift]: same constant comments as stated before

File: Cornell Sun/Cornell Sun/SearchViewController.swift [hunk 3]
```diff
@@ -7,81 +7,265 @@
...
+    }
+
+    func emptyView(for listAdapter: ListAdapter) -> UIView? {
+        //eventually show uh-oh no searches found
+        return emptyLoadingView
+    }
+
+    func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        let distance = scrollView.contentSize.height - (targetContentOffset.pointee.y + scrollView.bounds.height)
+        if !loading && distance < 300 && !collectionView.isHidden && !endOfResults {
```
Comment (cdsciavolino) [inline Cornell Sun/Cornell Sun/SearchViewController.swift]: you could consider putting 300 into a constant variable too

File: Cornell Sun/Cornell Sun/Search/SearchSectionController.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+class SearchSectionController: ListSectionController {
+
+    var entry: PostObject!
+    weak var delegate: TabBarViewControllerDelegate?
+
+    override init() {
+        super.init()
+        inset = UIEdgeInsets(top: 0, left: 0, bottom: 15, right: 0)
```
[the remark this answers is not in the document]
Comment (AAAstorga) [reply Cornell Sun/Cornell Sun/Search/SearchSectionController.swift]: No, it is so it makes the space between collection view cells

File: Cornell Sun/Extensions.swift [hunk 5]
```diff
@@ -8,17 +8,14 @@
 
 import UIKit
 import Kingfisher
+import Realm
+import RealmSwift
 
 func captionMaxHeight(width: CGFloat) -> CGFloat {
      let tmpStr = "A rink attendant collect fish thrown by the Cornell students in a tradition that spans over decades. (Cameron Pollack/ Sun Photography Editor)"
```
[the remark this answers is not in the document]
Comment (AAAstorga) [reply Cornell Sun/Extensions.swift]: Realm is how we do persistence for bookmarks and such

[the remark this answers is not in the document]
Comment (mindylou) [reply Cornell Sun/Extensions.swift]: wait I meant `captionMaxHeight` lol

[the remark this answers is not in the document]
Comment (AAAstorga) [reply Cornell Sun/Extensions.swift]: oh lolz my b. It is used for caption height. like that string produces an optimal height number no matter what device. I can explain better in person

File: Cornell Sun/Cornell Sun/Search/SearchSectionController.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    override func cellForItem(at index: Int) -> UICollectionViewCell {
+        guard let cellForItemIndex = searchCellType(rawValue: index) else {
+            return UICollectionViewCell()
+        }
+
+        switch cellForItemIndex {
+        case .categoryCell:
```
[the remark this answers is not in the document]
Comment (AAAstorga) [reply Cornell Sun/Cornell Sun/Search/SearchSectionController.swift]: unfortunately I can't bc of all the different types of cells

**Pull request merged**